### PR TITLE
Introduce a new env variable to allow spaces in home path

### LINKED ISSUE #1320: Improve docs for resolving `space in home path` warning
**How would this feature be useful?**

The existing warning and docs are not enough for users to resolve it.

**Describe the solution you'd like**

Dedicated docs on how to move the home path and a clear link to it in the warning, as well as specifying it's the `Pipx Home` that has the space in it.

**Describe alternatives you've considered**

Hope people that are confused just find the existing discussions.

### Discussed in https://github.com/pypa/pipx/discussions/1315

<div type='discussions-op-text'>

<sup>Originally posted by **homocomputeris** April  1, 2024</sup>
I get the message every time I run pipx.

- What is home path here?
- Where are the docs on this issue? Nothing can be found here: https://www.google.com/search?hl=en&q=%22home%20path%22%20site%3Ahttps%3A%2F%2Fpipx.pypa.io%2Flatest%2F
- How to fix or disable this warning?</div>

### PR DESCRIPTION
**Pull request by Gitznik**
- [x] I have added a news fragment under `changelog.d/` (if the patch affects the end users)

## Summary of changes
Introduces an environment variable to silence the warning for having a space in the pipx home path.

Improves the docs on how to troubleshoot the warning and move the pipx installation location in general. This makes [the discussion](https://github.com/pypa/pipx/discussions/1330) I created a while ago redundant.

Resolves #1320

## Test plan

Tested by running

```
nox -s tests-3.8 -- -k test_environment
```

### CODE REVIEW

**Comment by huxuan** [comment]
> @huxuan if you have the time, could you please verify that the windows example for moving the pipx install location actually works? Or maybe even translate it to powershell?

Gotcha, will take a look later.

File: docs/installation.md [hunk 1]
```diff
@@ -223,3 +223,62 @@ You can easily get your shell's tab completions working by following instruction
...
+```bash
+NEW_LOCATION=~
+cache_dir=$(pipx environment --value PIPX_VENV_CACHEDIR)
+logs_dir=$(pipx environment --value PIPX_LOG_DIR)
+trash_dir=$(pipx environment --value PIPX_TRASH_DIR)
+home_dir=$(pipx environment --value PIPX_HOME)
+rm -rf $cache_dir $logs_dir $trash_dir
+mv $home_dir $NEW_LOCATION
+pipx reinstall-all
+```
```
Comment (huxuan) [inline docs/installation.md:284]:
````suggestion
```powershell
$NEW_LOCATION = Join-Path $HOME '.local/pipx'
$cache_dir = pipx environment --value PIPX_VENV_CACHEDIR
$logs_dir = pipx environment --value PIPX_LOG_DIR
$trash_dir = pipx environment --value PIPX_TRASH_DIR
$home_dir = pipx environment --value PIPX_HOME

Remove-Item -Recurse -Force -ErrorAction SilentlyContinue $cache_dir, $logs_dir, $trash_dir

Move-Item -Path $home_dir -Destination $NEW_LOCATION
pipx reinstall-all
```
````

BTW, the process works fine on my PC.

File: docs/installation.md [hunk 2]
```diff
@@ -223,3 +223,62 @@ You can easily get your shell's tab completions working by following instruction
...
+rm -rf $cache_dir $logs_dir $trash_dir
+mkdir -p $NEW_LOCATION && mv $home_dir $NEW_LOCATION
+pipx reinstall-all
+```
+
+### Windows
+
+Requires bash (for example `git bash`)
+
+Current default location: `~`
```
Comment (huxuan) [inline docs/installation.md:273]:
```suggestion
Current default location: `~/pipx`
```

According to the implementation here [1], the default location should be "~/pipx".

[1] https://github.com/pypa/pipx/blob/dc9f9a849a918c68bc2cc99c72651022371fcf9f/src/pipx/paths.py#L16

**Comment by huxuan** [comment]
> Could you please verify that the windows example for moving the pipx install location actually works? Or maybe even translate it to powershell?

Verified and translated! PTAL

File: docs/installation.md (hunk 1 above)
Comment (Gitznik) [reply docs/installation.md:284]: In powershell, the `Move-Item` command moves the files in the specified paths into the `-Destination` path?

Comment (huxuan) [reply docs/installation.md:284]: Good catch! I asked ChatGPT and tested on my PC, it seems to depend on the existence of destination directory.

If the destination directory exists, the behavior is just like bash. When the destionation does not exists, it works more like a rename. That is why it works on my side previously.

So maybe we should remove the destionation directory before `Move-Item` to ensure the rename behavior. WDYT?

BTW, I updated the suggested changes accordingly.

Comment (Gitznik) [reply docs/installation.md:284]: Gotcha, yeah nice this looks good to me. Thanks for helping me out here, I'm useless when it comes to windows ;)

**Review by huxuan — DISMISSED** [review]

**Comment by Gitznik** [comment]
Sorry had to add just one small change to make sure spaces in the paths are not a problem when using bash.

**Comment by huxuan** [comment]
> Sorry had to add just one small change to make sure spaces in the paths are not a problem when using bash.

looks like the double quotes are also needed for the dirs and $NEW_LOCATION in powershell.

**Review by huxuan — DISMISSED** [review]

File: docs/installation.md [hunk 3]
```diff
@@ -257,3 +257,71 @@ You can easily get your shell's tab completions working by following instruction
...
+logs_dir=$(pipx environment --value PIPX_LOG_DIR)
+trash_dir=$(pipx environment --value PIPX_TRASH_DIR)
+home_dir=$(pipx environment --value PIPX_HOME)
+# If you wish another location, replace the if block below
+# and set `NEW_LOCATION` explicitly
+if [ -n "$XDG_DATA_HOME" ]; then
+    NEW_LOCATION="$XDG_DATA_HOME"
+else
+    NEW_LOCATION="$HOME/.local/share"
+fi
```
Comment (chrysle) [inline docs/installation.md:297]: We might be able to simplify this:

```suggestion
# If you wish another location, replace the expression below
# and set `NEW_LOCATION` explicitly
NEW_LOCATION="${XDG_DATA_HOME:-$HOME/.local/share}"
```

File: docs/installation.md [hunk 4]
```diff
@@ -257,3 +257,71 @@ You can easily get your shell's tab completions working by following instruction
...
+rm -rf "$cache_dir" "$logs_dir" "$trash_dir"
+mkdir -p $NEW_LOCATION && mv "$home_dir" $NEW_LOCATION
+pipx reinstall-all
+```
+
+### Windows
+
+Requires bash (for example `git bash`)
```
Comment (chrysle) [inline docs/installation.md:305]:
```suggestion
Requires bash (for example `git bash`).
```

File: docs/installation.md [hunk 5]
```diff
@@ -257,3 +257,71 @@ You can easily get your shell's tab completions working by following instruction
...
+Remove-Item -Recurse -Force -ErrorAction SilentlyContinue "$cache_dir", "$logs_dir", "$trash_dir"
+
+# Remove the destination directory to ensure rename behavior of `Move-Item`
+Remove-Item -Recurse -Force -ErrorAction SilentlyContinue "$NEW_LOCATION"
+
+Move-Item -Path $home_dir -Destination "$NEW_LOCATION"
+pipx reinstall-all
+```
+
+If you would prefer doing it in bash via git-bash/WSL, feel free to use
```
Comment (chrysle) [inline docs/installation.md:325]: Don't the instructions say that git-bash is required, above?

File: docs/troubleshooting.md [hunk 6]
```diff
@@ -148,13 +148,28 @@ In Pipx version 1.5.0, this was reverted for Windows and MacOS. It defaults agai
...
+automatically) and then reinstall all packages.
 
-```
-rm -rf ~/.local/pipx/{.cache,logs,trash}
-mkdir -p ~/.local/share && mv ~/.local/pipx ~/.local/share/
-pipx reinstall-all
-```
+Please refer to [Installation](installation.md#moving-your-pipx-installation) on how to move it.
+
+## Warning: Found a space in the pipx home path
```
Comment (chrysle) [inline docs/troubleshooting.md:155]: Since this is documented here, now, we should replace the link pointing to the GH discussion at [this paragraph](https://pipx--1446.org.readthedocs.build/1446/installation/#installation-options).

File: src/pipx/animate.py [hunk 7]
```diff
@@ -4,8 +4,8 @@
 from threading import Event, Thread
 from typing import Generator, List
 
+import pipx.emojis as emojis
```
Comment (chrysle) [inline src/pipx/animate.py:7]: This change seems unrelated... I prefer to not mix changes, as it gets unclear why they were introduced. Also, is it necessary?

File: src/pipx/emojis.py [hunk 8]
```diff
@@ -1,15 +1,7 @@
 import os
 import sys
 
-
-def strtobool(val: str) -> bool:
```
Comment (chrysle) [inline src/pipx/emojis.py:5]: Same for this change. Could you bundle them in a separate PR?

File: docs/troubleshooting.md [hunk 9]
```diff
@@ -148,13 +148,28 @@ In Pipx version 1.5.0, this was reverted for Windows and MacOS. It defaults agai
...
+
+If you want to use a script installed via pipx in a shebang itself (common for example for the aws cli), you run into a similar problem, as the path to the installed script will contain a space.
+
+### How to fix
+
+You can generally fix this by using our default locations, as long as your `$HOME` path does not contain spaces.
+Please refer to our [Installation](installation.md#moving-your-pipx-installation) docs on how to move the `pipx` installation.
 
-For moving the paths back after 1.5.0, you can perform the same steps, switching the paths around.
+If you're really sure you want to stick to your path with spaces, you can set the `PIPX_HOME_ALLOW_SPACE` environment variable to `true`.
```
Comment (chrysle) [inline docs/troubleshooting.md:175]:
```suggestion
If you're really sure you want to stick to your path with spaces, to suppress the warning set the `PIPX_HOME_ALLOW_SPACE` environment variable to `true`.
```

File: src/pipx/emojis.py (hunk 8 above)
Comment (Gitznik) [reply src/pipx/emojis.py:5]: It felt wrong using `strtobool` from `emojis.py` in `paths.py` for something completely emoji unrelated, so I moved it to utils. This required the import changes to fix a circular import.
Do you think I should stick to importing it from emojis for this PR?

Comment (chrysle) [reply src/pipx/emojis.py:5]: Probably, yes; this and the circular import problem mentioned in #1333 could be resolved in one go.

File: docs/installation.md (hunk 5 above)
Comment (Gitznik) [reply docs/installation.md:319]: This was a leftover from what I copied from the discussion. Removed it.

File: docs/troubleshooting.md (hunk 6 above)
Comment (Gitznik) [reply docs/troubleshooting.md:155]: 100%

File: src/pipx/emojis.py (hunk 8 above)
Comment (Gitznik) [reply src/pipx/emojis.py]: Removed it from this PR

**Review by chrysle — APPROVED** [review]
